### ISSUE TITLE
Get arguments from a scoreboard?

### ISSUE THREAD

**Issue by mcgamerguy** [gibbsly/gm#17]
I was wondering if it would somehow be posible to instead of setting the x in the arguments, have it be set from a scoreboard, as i am trying to use the sin and cos for a lot of different values. Could this get added? For example using another argument you can set to true and then x becomes the name of the scoreboard instead of a value?

**Comment by GlitchlessCode** [comment]
I have a few issues with this idea (not that it's a bad idea, but just a few things to think about).

1. Assuming we leave the underlying functionality of each of the mathematical functions the same, still using data storage and whatnot in order to perform each specific operation using actual floating point numbers, then adding the underlying system to use the scoreboard values, as suggested, would simply copy the scoreboard values to the data storage, which is exactly what you'd need to do right now anyways. This means that the overhead of doing this is theoretically the same as just doing it yourself.
2. On top of point 1, it would not only simply act as a shift of overhead from the user to the library, but would also add additional overhead in the form of needing to check if the "other argument" was set, so the function would know whether to pull from a scoreboard value or from the direct macro value. This is would unfortunately quite complex, due to the hacky nature of macros. (If you want an idea of how this would work, I'm willing to share my thoughts, this is just a superficial overview)
3. Point 2 could be circumvented by adding a specialized version of each functions specifically for use with scoreboards, instead of using the proposed "other argument". However, this would require a reorganization of the file structure, to allow for one set of macro input functions, and another set of scoreboard input functions. 

In general, I do think that looking into this further could be an interesting idea, but overall could just be more trouble than it's worth, however, I don't know. 

One other thing to consider is that this issue doesn't really clarify if it only wants scoreboard inputs, or also scoreboard outputs, as that would be another thing to consider (though in that case, one would likely want to also address what format to expect, whether it would just be scaled to 1, or upscaled to a fixed decimal point system). Evidently, this is out of scope of the upfront face value of the issue at hand, but is still important to consider.

Overall, I think this issue could be something that is addressed, but at least as of right now, the way to get the least overhead, and just work, is to leave it to the user of the library to copy the scoreboard value to a data storage, and just use that storage as the macro input.

Obviously this isn't set in stone, these are just my thoughts on it though.

**Comment by mcgamerguy** [comment]
The issue im having with the last thing you said is this:
I set my scoreboard to 1
/execute store result storage gm:io x double 1 run scoreboard players get @s x
/function gm:cos with storage gm:io x
then i get the error: "Invalid argument type: DOUBLE, expected Compound"

**Comment by mcgamerguy** [comment]
oeps, didnt mean to close it, mb

**Comment by GlitchlessCode** [comment]
> oeps, didnt mean to close it, mb

XD I've done that before, all good

**Comment by GlitchlessCode** [comment]
> The issue im having with the last thing you said is this: I set my scoreboard to 1 /execute store result storage gm:io x double 1 run scoreboard players get @s x /function gm:cos with storage gm:io x then i get the error: "Invalid argument type: DOUBLE, expected Compound"

In this case, just use 
```mcfunction
function gm:cos with storage gm:io
```
instead of 
```mcfunction
function gm:cos with storage gm:io x
```
This is because the `with` syntax expects a compound object, similar to if you were feeding the macro manually. For instance, manually, you'd do:
```mcfunction
function gm:cos {x:1.0d}
```
but by using the syntax you put, instead of getting
```json
{
  "x":1.0
}
```
the macro replacer just gets
```json
1.0
```
because you're pulling telling it to read the contents of the theoretical compound object at `x`, where it finds a double instead of a compound. Instead, you want it to read the compound object that forms the entire data storage `gm:io`, and then it'll read the `x` property from that compound object.

Hopefully that works! Let me know if you need any more help.

**Comment by mcgamerguy** [comment]
ohh, great ty! ill test that out

**Comment by mcgamerguy** [comment]
Yeah that does work, also i was trying to use some calculations so summon a fireball which is moving in the direction the player is looking, but it seems like the sin and cos functions isnt exact enough for it to work

**Comment by gibbsly** [comment]
You don't need sin/cos for doing that, you can apply motion based on look angle with 2 commands, 3 if you don't have a fixed entity already summoned

you can do something like `execute summon fireball run function apply_motion`
#### function `apply_motion`
```
summon area_effect_cloud ~ ~ ~ {UUID:[I;0,0,0,15]}
execute positioned 0.0 0.0 0.0 run tp 0-0-0-0-f ^ ^ ^1
data modify entity @s Motion set from entity 0-0-0-0-f Pos
```

**Comment by mcgamerguy** [comment]
yeah i also found this, but when i used that on a snowball there is a small glitch of it first falling downwards for everyone, then tping to where it was suppost to go, this is from 1.21.1 btw, any idea on how to fix it?
